### PR TITLE
Introducing findable regions for testing

### PR DESCRIPTION
**Pull request by taras**
## Motivation

I described in https://github.com/vadimdemedes/ink-testing-library/issues/10 the motivation for exposing Ink's DOM like API to make it easier to assert on specific components props rather than parsing a potentially complex multiline string. I created that issue before I had a lot of experience with the Ink test suite. 

Ink test suite it distinctly *big*, meaning that it tests the entire system by asserting on output rather than any internal details of the system. For this reason, my request to expose the DOM API felt like a compromise. 

Based on @cowboyd's suggestion, I spent some time looking at how we could maintain the whole system testing approach while giving us a way to identify specific areas of the output. I started to read [XTerm Control Sequences](https://www.xfree86.org/current/ctlseqs.html) in hopes of finding some character that terminals ignored by would allow me to find specific strings in output. In the XTerm Control Sequences document, I found Application Program-Control functions with the following description:

> APC Pt ST: xterm implements no APC functions; P t is ignored. P t need not be printable characters

APC fits the bill well because it has a named sequence that we can be used to distinguish one APC from another. APC seems to be a perfect fit for our purpose. The only reference that I found to APCs on Google was in xterm.js repository in an issue called [Consider supporting adding custom APC handlers, similar to CSI/OSC handlers](https://github.com/xtermjs/xterm.js/issues/2300) 

According to @jerch in https://github.com/xtermjs/xterm.js/issues/2300#issuecomment-510457550, 

> Currently APC and PM realms are not specified by any official document, there is no prior art to learn from besides the global ignore rule (which almost every VT parser applies) 

The message above seems to reinforce the idea that APC is a fair game because terminals ignore them. 

## Approach

This PR introduces a concept I called a region. A region starts with a named APC and ends with same-named APC prefixed with /, similar to how XML tags open with `<name>`and end with `</name>`. For example, `\u001B_{name}\u001B\\{text}\u001B_/{name}\u001B\\`.

I attempted to spike using unstable__transformChildren but it didn't work as expected because unstable__transformChildren wraps each text element. I needed the ability to wrap all children with one set of opening and closing tags.

I introduced a new unstable attribute called `unstable__regionName`. I applied this attribute to Box, Static and Text components which allows me to say `<Text name="message">Hello World</Text>`. This allows me to identify this region in output using `findByName` testing helper. This helper allows finding nested sequences. For example, 

```jsx
<Static name="static">
    <Text name="line">
        <Color red>Hello</Color>
    </Text>
    <Text name="line">
        <Color blue>World</Color>
    </Text>
</Static>
 ```

`findByName('static line')` will return `[red("Hello"), blue("World")]`

Included some tests to show the cases that I covered by the helper implementation.

## TODO

- [ ] Move findByName helper into ink-testing-library

### CODE REVIEW

**Comment by jerch** [comment]
Well the truth about APC is abit more complicated. Looking at ECMA-48 shows that APC belongs to so called "control strings" - sequence types, that are meant to control certain endpoints:
- APC: application program command
- DCS: device control string
- OSC: operating system command
- PM: privacy message
- SOS: start of string

Those string command are identified by certain introducers, and end with ST normally. Beside the allowed payload of mostly ASCII printables, they are not further specced out by the standard. This was done on purpose, they should leave room for later addons or custom functionality by implementors.
DEC was one of those implementors with its VT line (and printers). It made heavy usage of DCS commands with its own format to drive certain aspects of terminals and printers. Late devices also made use of OSC commands, while APC/SOS/PM was never used by DEC itself. All devices would silently swallow those unknown commands.
Most commonly used terminal emulators these days claim to be VT compatible. But be aware, many are not in regards of rare sequence types. We have several emulators that evolved from perceivable behavior of the devices, while specs were not available in public or way to expensive. Their parsers typically dont know anything about APC or DCS, and will treat the data as normal printable (dumping it to screen). OSC is in a better shape across the board (since more common in emulators), but again might fail for several reason like payload length restrictions.

**TL;DR**: As sad as the situation is - dont use APC/DCS/OSC/SOS/PM as a general purpose mechanism, to hide or blackout data for terminal emulators without proper testing beforehand (sequence type supported at all? payload limits?). Some attempts might work on most emulators, others might fail miserably on most. In general the support by emulators is somewhere along OSC > DCS >> APC (some use private APC, I am not aware of any implementing PM/SOS support).

May I ask what your goal for using APC is? Taking your example above `\u001B_{name}\u001B[{text}\u001B_/{name}\u001B[` means in terms of terminal sequences (in a DEC compatible way, which slightly differs from EMCA-48, see also [VT500 parser](https://vt100.net/emu/dec_ansi_parser)):
- APC_start
- APC_payload: `{name}`
- APC_abort (here the next ESC symbol will abort the APC state)
- `CSI {` (defined as DECSERA)
- PRINT `text}`
- APC_start
- APC_payload `/{name}`
- APC_abort (again by following ESC, which stays meaningful in APC)
- CSI start (not finished)

I am not sure, what you were after with this example, but prolly not that list of commands?

> .. I started to read XTerm Control Sequences in hopes of finding some character that terminals ignored by would allow me to find specific strings in output. ..

If your main goal is to embed some kind markup to identify regions in the output ignored by most terminals - well you might see better support without screwing up the output with a custom OSC command (e.g. `OSC 12345 ; <some payload> BEL`) Here the number is meant as function identifier - if an emulator does not find that particular function is should swallow the content silently. Again a certain emulator might have payload length restrictions in place (needs testing again).

**Comment by taras** [comment]
mindblown

@jerch, thank you for jumping in and the thorough response to my message. I appreciate your insight because it's surprisingly challenging to find this information online. Maybe I'm looking in the wrong place, but having it all summarized is invaluable.

> May I ask what your goal for using APC is?

We need an easy way to reliably retrieve text from specific areas on the screen in tests. We can suppress these sequences in regular runtime because they offer no value outside of testing.

> \u001B_{name}\u001B[{text}\u001B_/{name}\u001B[

The above example is misleading because it's a combination pseudo template and actual sequence output. The actual output would look like this: `\u001B_message\u001B[Hello World\u001B_/message\u001B[`. It provides similar boundaries XML: `<message>Hello World</message>`

> Again a certain emulator might have payload length restrictions in place (needs testing again)

What would be an example of such an emulator?

> Again a certain emulator might have payload length restrictions in place (needs testing again).

Ink allows building CLIs with node.js. I don't know in what environments node.js CLI tools are used, so it's difficult to tell what emulators we should test against. Can you make any recommendations on how to narrow down our targets?

**Comment by jerch** [comment]
> Maybe I'm looking in the wrong place, but having it all summarized is invaluable.

Yes the documentation is abit thin and scattered, seems terminals almost got forgotten during the last 2 decades, but lately see more attention again.
 
Best old sources regarding this (late 80s / early 90s):
- [ECMA-48](https://www.ecma-international.org/publications/files/ECMA-ST/Ecma-048.pdf)
- [DEC STD 070](https://archive.org/embed/bitsavers_decstandar0VideoSystemsReferenceManualDec91_74264381)
- [VT100.net](https://vt100.net/)


The latter two are DEC related, which are more relevant in terms of emulators used as of today on unix like systems. ECMA-48 is the only document we have as an official standard regarding sequence types, but outdated itself. It is still good to get the basic picture, along with other specifications like ECMA-35 (again somewhat outdated). To get closer to the truth as of what we see today in emulators, you might also look at changelogs / docs of `xterm`, which kept progressing and adding important features (like Unicode/UTF-8 support, palette colors, SIXEL etc.) and became to the de-facto emulation standard.

> The above example is misleading because it's a combination pseudo template and actual sequence output. The actual output would look like this: `\u001B_message\u001B[Hello World\u001B_/message\u001B[`. It provides similar boundaries XML: `<message>Hello World</message>`

Maybe this is just a typo on your side - the official APC terminator is ST, which is `\x1b \` in 7-bit or `\x9c` in 8-bit. Written as string literals the `\` has to be escaped in most languages (in JS as `'\x1b\\'` or `'\u001b\\'`). You terminate the APC command by a `\x1b[` which is the CSI introducer, which would "eat" any following printable char to form a full CSI sequence (here `CSI H` defined as CUP). Maybe you meant to write `\u001B_message\u001B\\Hello World\u001B_/message\u001B\\`?

> What would be an example of such an emulator?

`rxvt-unicode` will break an OSC command after ~32kB of payload. You can test this in any emulator like this:
```bash
echo -n "Hello \x1b]12345;"; for i in {1..1000}; do echo -n =; done; echo "\x07World"
```
which starts showing excessive '=' around 33000. Btw `vte` seems to output that payload always prolly not ignoring unknown sequences at all. You can test it likewise with APC and will find, that most emulators do not handle APC well (always showing the payload).

> We need an easy way to reliably retrieve text from specific areas on the screen in tests. We can suppress these sequences in regular runtime because they offer no value outside of testing.

Does the test scenario involve emulator interactions at all? If not then APC support is more of academical interest (and ofc, in general APC is the better choice for your needs), otherwise OSC might give better results.

**Comment by taras** [comment]
> You terminate the APC command by a \x1b[ which is the CSI introducer, which would "eat" any following printable char to form a full CSI sequence (here CSI H defined as CUP).

This would explain why I would sometimes see a missing character after the terminator and beginning of text. 

> Maybe you meant to write \u001B_message\u001B\\Hello World\u001B_/message\u001B\\

Yes, I corrected the implementation. Thank you so much for pointing it out. I would have spent a long time chasing my tail with that one.

> Does the test scenario involve emulator interactions at all?

No, it'll be strictly for capturing output and I'm planning to remove them in regular runtime.

**Comment by taras** [comment]
@vadimdemedes I decided to abandon this spike after trying to get slice-ansi to properly parse these APCs. It would require changes that increase the complexity of the implementation of slice-ansi for scenarios that might never occur again. It's unlikely that changes to dependencies would be merged to support this, hence there is no point to proceed with this approach. 

@jerch I'm immensely grateful for your feedback in this branch. I learned a lot more about terminal than I ever knew was there. I hope Google picks up this threads and others will find it useful.
